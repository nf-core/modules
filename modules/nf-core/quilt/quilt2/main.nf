process QUILT_QUILT2 {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/r-quilt:2.0.1--r44h503566f_1':
        'biocontainers/r-quilt:2.0.1--r44h503566f_1' }"

    input:
    tuple val(meta), path(bams), path(bais), path(bamlist), path(reference_vcf_file), path(reference_vcf_file_index), path(posfile), path(phasefile), val(chr), val(regions_start), val(regions_end), val(ngen), val(buffer), path(genetic_map_file)
    tuple val(meta2), path(fasta)

    output:
    tuple val(meta), path("*.vcf.gz"),              emit: vcf
    tuple val(meta), path("*.vcf.gz.tbi"),          emit: tbi,   optional:true
    tuple val(meta), path("RData", type: "dir"),    emit: rdata, optional:true
    tuple val(meta), path("plots", type: "dir"),    emit: plots, optional:true
    path "versions.yml",                            emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args                        =   task.ext.args ?: ''
    def _prefix                     =   task.ext.prefix ?: "${meta.id}"
    def extensions                  =   bams.collect { it.extension }
    def extension                   =   extensions.flatten().unique()
    def list_command                =   extension == ["bam"]  ? "--bamlist="                       :
                                        extension == ["cram"] ? "--reference=${fasta} --cramlist=" : ""
    def genetic_map_file_command    =   genetic_map_file      ? "--genetic_map_file=${genetic_map_file}"     : ""
    def posfile_command             =   posfile               ? "--posfile=${posfile}"                       : ""
    def phasefile_command           =   phasefile             ? "--phasefile=${phasefile}"                   : ""
    if (!(args ==~ /.*--seed.*/)) {args += " --seed=1"}

    """
    if [ -n "$bamlist" ] ;
    then
        BAM_LIST="$bamlist"
    else
        printf "%s\\n" $bams | tr -d '[],' > all_files.txt
        BAM_LIST="all_files.txt"
    fi

    QUILT2.R \\
        ${list_command}\$BAM_LIST \\
        $genetic_map_file_command \\
        $posfile_command \\
        $phasefile_command \\
        --chr=$chr \\
        --regionStart=$regions_start \\
        --regionEnd=$regions_end \\
        --nGen=$ngen \\
        --buffer=$buffer \\
        --nCores=$task.cpus \\
        --outputdir="." \\
        --reference_vcf_file=$reference_vcf_file \\
        $args


    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        r-base: \$(Rscript -e "cat(strsplit(R.version[['version.string']], ' ')[[1]][3])")
        r-quilt2: \$(Rscript -e "cat(as.character(utils::packageVersion(\\"QUILT2\\")))")
    END_VERSIONS
    """
}
