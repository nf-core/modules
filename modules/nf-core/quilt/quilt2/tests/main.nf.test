nextflow_process {

    name "Test Process QUILT2"
    script "../main.nf"
    config "./nextflow.config"
    process "QUILT_QUILT2"

    tag "modules"
    tag "modules_nfcore"
    tag "quilt/quilt2"
    tag "quilt2"

    setup {
        run("GAWK") {
            script "../../../gawk/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'map' ],
                    file(params.modules_testdata_base_path + "genomics/homo_sapiens/genome/genome.GRCh38.eagle.22.map.gz", checkIfExists: true)
                ])
                input[1] = Channel.of('{ OFS=" " } { if (\$2 == "15333045") { \$3 = "0.8914971326" } ; print \$2, \$3, \$4 }').collectFile(name:"program.txt")
                input[2] = false
                """
            }
        }
    }

    test("QUILT2 no optional files") {
        tag "test"
        when {
            params {
                quilt_args = "--seed=1"
            }
            process {
                """
                input[0] = Channel.of([
                    [ id:'test', chr:'chr22' ],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/NA12878.chr22.1X.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/NA12878.chr22.1X.bam.bai', checkIfExists: true),
                    [],
                    file(params.modules_testdata_base_path + "genomics/homo_sapiens/popgen/1000GP.chr22.vcf.gz", checkIfExists: true),
                    file(params.modules_testdata_base_path + "genomics/homo_sapiens/popgen/1000GP.chr22.vcf.gz.csi", checkIfExists: true),
                    [], [],
                    "chr22", "16570000", "16610000", "100", "10000"
                ]).combine(
                    GAWK.out.output.map{it[1]}
                )
                input[1] = [[id: 'GRCh38'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("QUILT2 optional output") {
        when {
            params {
                quilt_args = "--save_prepared_reference=TRUE --make_plots=TRUE --seed=1"
            }
            process {
                """
                input[0] = Channel.of([
                    [ id:'test', chr:'chr22' ],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/NA12878.chr22.1X.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/NA12878.chr22.1X.bam.bai', checkIfExists: true),
                    [],
                    file(params.modules_testdata_base_path + "genomics/homo_sapiens/popgen/1000GP.chr22.vcf.gz", checkIfExists: true),
                    file(params.modules_testdata_base_path + "genomics/homo_sapiens/popgen/1000GP.chr22.vcf.gz.csi", checkIfExists: true),
                    [], [],
                    "chr22", "16570000", "16610000", "100", "10000"
                ]).combine(
                    GAWK.out.output.map{it[1]}
                )
                input[1] = [[id: 'GRCh38'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)]
                """
            }
        }

        then {
            def dir = new File(process.out.plots[0][1])
            def list = []
            dir.eachFileRecurse { file -> list << file.getName() }
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.vcf + process.out.tbi +
                    list.sort() +
                    process.out.rdata + process.out.versions
                ).match() }
            )
        }

    }

    test("QUILT with bamlist") {
        when {
            params {
                quilt_args = "--seed=1"
            }
            process {
                """
                bamlist = Channel.of(
                    "NA12878.bam"
                ).collectFile(name : 'bamlist.txt', newLine : true)

                ch_input = Channel.of([
                    [ id:'test', chr:'chr22' ],
                    file(params.modules_testdata_base_path + "genomics/homo_sapiens/illumina/bam/NA12878.chr22.1X.bam", checkIfExists: true),
                    file(params.modules_testdata_base_path + "genomics/homo_sapiens/illumina/bam/NA12878.chr22.1X.bam.bai", checkIfExists: true),
                ]).combine(
                    bamlist
                ).combine(Channel.of([
                    file(params.modules_testdata_base_path + "genomics/homo_sapiens/popgen/1000GP.chr22.vcf.gz", checkIfExists: true),
                    file(params.modules_testdata_base_path + "genomics/homo_sapiens/popgen/1000GP.chr22.vcf.gz.csi", checkIfExists: true),
                    [], [],
                    "chr22", "16580000", "16600000", "100", "20000"
                ])).combine(
                    GAWK.out.output.map{it[1]}
                )
                input[1] = Channel.of([
                    [id:'refHG37_chr22'],
                    file(params.modules_testdata_base_path + "genomics/homo_sapiens/genome/GRCh37_chr22.fasta.gz", checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }
}