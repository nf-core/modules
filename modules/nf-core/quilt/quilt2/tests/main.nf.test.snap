{
    "QUILT2 no optional files": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    
                ],
                "plots": [
                    
                ],
                "rdata": [
                    
                ],
                "tbi": [
                    
                ],
                "vcf": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-29T14:56:04.174405471"
    }
}